wallet-next/src/sync.rs

```diff
@@ -0,0 +1,93 @@
+};
+use penumbra_proto::Protobuf;
+use penumbra_transaction::action::output;
+
+/// Corresponds to a row in the notes table
+pub struct NoteRecord {
+    pub note_commitment: Vec<u8>,
+    pub height_spent: Option<u64>, //None if unspent, otherwise spent at height_spent
+    pub height_created: u64,
+    // note contents themselves:
+    pub diversifier: Vec<u8>,
+    pub amount: u64,
+    pub asset_id: Vec<u8>,
+    pub transmission_key: Vec<u8>,
+    pub blinding_factor: Vec<u8>,
+    // precomputed decryption of the diversifier
+    pub diversifier_index: Vec<u8>,
+    // the nullifier for this note, used to detect when it is spent
+    pub nullifier: Vec<u8>,
+}
+
+/// Contains the results of scanning a single block.
+struct ScanResult {
+    // write as new rows
+    pub new_notes: Vec<NoteRecord>,
+    // use to update existing rows
+    pub spent_nullifiers: Vec<Nullifier>,
+    pub height: u64,
+}
+
+#[tracing::instrument(skip(fvk, outputs, nullifiers))]
+fn scan_block(
+    fvk: &FullViewingKey,
+    note_commitment_tree: &mut penumbra_crypto::merkle::NoteCommitmentTree,
+    CompactBlock {
+        height,
+        outputs,
+        nullifiers,
+    }: CompactBlock,
+) -> ScanResult {
+    let new_notes: Vec<NoteRecord>;
+
+    for output::Body {
+        note_commitment,
+        ephemeral_key,
+        encrypted_note,
+    } in outputs.into_iter()
+    {
+        // Unconditionally insert the note commitment into the merkle tree
+        tracing::debug!(?note_commitment, "appending to note commitment tree");
+        note_commitment_tree.append(&note_commitment);
+
+        // Try to decrypt the encrypted note using the ephemeral key and persistent incoming
+        // viewing key -- if it doesn't decrypt, it wasn't meant for us.
+        if let Ok(note) = Note::decrypt(encrypted_note.as_ref(), fvk.incoming(), &ephemeral_key) {
+            tracing::debug!(?note_commitment, ?note, "found note while scanning");
+            // Mark the most-recently-inserted note commitment (the one corresponding to this
+            // note) as worth keeping track of, because it's ours
+            note_commitment_tree.witness();
+
+            // Insert the note associated with its computed nullifier into the nullifier map
+            let (pos, _auth_path) = note_commitment_tree
+                .authentication_path(&note_commitment)
+                .expect("we just witnessed this commitment");
+
+            let nullifier = fvk.derive_nullifier(pos, &note_commitment);
+
+            let record = NoteRecord {
+                note_commitment: note_commitment.encode_to_vec(),
+                height_spent: None,
+                height_created: height,
+                diversifier: note.diversifier().0.to_vec(),
+                amount: note.amount(),
+                asset_id: note.asset_id().encode_to_vec(),
+                transmission_key: note.transmission_key().0.to_vec(),
+                blinding_factor: note.note_blinding().0.to_vec(),
```

Comment:
This code will get cleaner once the `NoteRecord` is a domain type that contains a `Note`, since the note can just be used directly.

xref https://github.com/penumbra-zone/penumbra/pull/667/files#r862422793